clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -605,7 +606,9 @@ public void abortTransaction() throws ProducerFencedException {
      * Implementation of asynchronously send a record to a topic.
      */
     private Future<RecordMetadata> doSend(ProducerRecord<K, V> record, Callback callback) {
-        ensureProperTransactionalState();
+        if (transactionManager != null)
```

Comment:
I agree it's subjective. The reason I preferred moving the check here is that it's easier to dismiss the function if you know you are not concerned about the transactional/idempotent cases. Otherwise, you have to descend into it to find that it just returns. My preference is not too strong either way though.